# Schema Validation — Customers

## 1. Primary Key Validation

### Question
Is `customer_id` a valid primary key?

We need to check:
- total number of rows
- number of unique `customer_id` values
- number of null `customer_id` values

In [5]:
import pandas as pd

df = pd.read_csv("../data/olist_customers_dataset.csv")

df.head(10)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP
5,879864dab9bc3047522c92c82e1212b8,4c93744516667ad3b8f1fb645a3116a4,89254,jaragua do sul,SC
6,fd826e7cf63160e536e0908c76c3f441,addec96d2e059c80c30fe6871d30d177,4534,sao paulo,SP
7,5e274e7a0c3809e14aba7ad5aae0d407,57b2a98a409812fe9618067b6b8ebe4f,35182,timoteo,MG
8,5adf08e34b2e993982a47070956c5c65,1175e95fb47ddff9de6b2b06188f7e0d,81560,curitiba,PR
9,4b7139f34592b3a31687243a302fa75b,9afe194fb833f79e300e37e580171f22,30575,belo horizonte,MG


In [6]:
df.shape[0]

99441

In [8]:
df["customer_id"].nunique()

99441

In [15]:
df["customer_id"].isna().sum()

np.int64(0)

### Insight

`customer_id` contains 99,441 unique and non-null values across 99,441
records. Therefore, it satisfies the uniqueness and non-null requirements
for a primary key in the `customers` table.

### Question

Is `customer_unique_id` a unique identifier for each customer in the customers table, or can the same customer appear in multiple records?

we need to check: 
- total number of rows
- number of unique `customer_unique_id` values
- number of null `customer_unique_id` values

In [18]:
df.shape[0]

99441

In [19]:
df["customer_unique_id"].nunique()

96096

In [21]:
df["customer_unique_id"].isna().sum()

np.int64(0)

### Insight

`customer_id` is unique for every record, while `customer_unique_id` is
not unique: there are 99,441 records but only 96,096 unique
`customer_unique_id` values.

This suggests that `customer_id` identifies a customer record, whereas
`customer_unique_id` represents a customer at the business level and may
appear in multiple records.

The relationship between repeated `customer_unique_id` values and
multiple orders will be validated later using the `orders` table.